crates/storage-pg/src/queue/worker.rs

```diff
@@ -0,0 +1,237 @@
+    type Error = DatabaseError;
+
+    #[tracing::instrument(
+        name = "db.queue_worker.register",
+        skip_all,
+        fields(
+            worker.id,
+            db.query.text,
+        ),
+        err,
+    )]
+    async fn register(
+        &mut self,
+        rng: &mut (dyn RngCore + Send),
+        clock: &dyn Clock,
+    ) -> Result<Worker, Self::Error> {
+        let now = clock.now();
+        let worker_id = Ulid::from_datetime_with_source(now.into(), rng);
+        tracing::Span::current().record("worker.id", tracing::field::display(worker_id));
+
+        sqlx::query!(
+            r#"
+                INSERT INTO queue_workers (queue_worker_id, registered_at, last_seen_at)
+                VALUES ($1, $2, $2)
+            "#,
+            Uuid::from(worker_id),
+            now,
+        )
+        .traced()
+        .execute(&mut *self.conn)
+        .await?;
+
+        Ok(Worker { id: worker_id })
+    }
+
+    #[tracing::instrument(
+        name = "db.queue_worker.heartbeat",
+        skip_all,
+        fields(
+            %worker.id,
+            db.query.text,
+        ),
+        err,
+    )]
+    async fn heartbeat(
+        &mut self,
+        clock: &dyn Clock,
+        worker: Worker,
+    ) -> Result<Worker, Self::Error> {
+        let now = clock.now();
+        let res = sqlx::query!(
+            r#"
+                UPDATE queue_workers
+                SET last_seen_at = $2
+                WHERE queue_worker_id = $1 AND shutdown_at IS NULL
+            "#,
+            Uuid::from(worker.id),
+            now,
+        )
+        .traced()
+        .execute(&mut *self.conn)
+        .await?;
+
+        // If no row was updated, the worker was shutdown so we return an error
+        DatabaseError::ensure_affected_rows(&res, 1)?;
+
+        Ok(worker)
+    }
+
+    #[tracing::instrument(
+        name = "db.queue_worker.shutdown",
+        skip_all,
+        fields(
+            %worker.id,
+            db.query.text,
+        ),
+        err,
+    )]
+    async fn shutdown(&mut self, clock: &dyn Clock, worker: Worker) -> Result<(), Self::Error> {
+        let now = clock.now();
+        let res = sqlx::query!(
+            r#"
+                UPDATE queue_workers
+                SET shutdown_at = $2
+                WHERE queue_worker_id = $1
+            "#,
+            Uuid::from(worker.id),
+            now,
+        )
+        .traced()
+        .execute(&mut *self.conn)
+        .await?;
+
+        DatabaseError::ensure_affected_rows(&res, 1)?;
+
+        Ok(())
+    }
+
+    #[tracing::instrument(
+        name = "db.queue_worker.shutdown_dead_workers",
+        skip_all,
+        fields(
+            db.query.text,
+        ),
+        err,
+    )]
+    async fn shutdown_dead_workers(
+        &mut self,
+        clock: &dyn Clock,
+        threshold: Duration,
+    ) -> Result<(), Self::Error> {
+        let now = clock.now();
```

Comment:
is it reasonable to rely on the system clock (which could drift between servers)?

I suppose we could use the Postgres database's clock alternatively. But I don't know which one is best, mostly just interested in considering it carefully

Reply:
That's a fair point, although I'd imagine multiple servers in the same datacenter usually have the same time source/NTP server?

The clock is abstracted through a trait though, so maybe at some point we can take into account time drift, and regularly sync the local system clock with the database or something, but I wouldn't worry too much about it for now

Reply:
It'd be good to document this at least maybe?

For a variety of reasons:

- assuming everyone will deploy workers only in the same datacentre is optimistic
  - besides, multi-datacentre or multi-zone is the hot stuff for high availability in case your datacentre does an OVH
- NTP could be misconfigured on some servers, or the network could be misconfigured preventing NTP from working
- one of the NTP servers may deviate from the others and may provide a fluke result to one server in isolation (probably a bit far-fetched and maybe NTP clients handle these problems themselves)

But overall I don't like the feeling of relying too heavily on the system clock for correctness.

Postgres' clock is at least probably consistent, assuming no multi-master Postgres fun and games (I don't know if that's feasible or not anyway).

If we can't get by just with locks and need to rely on some sort of clock, I have to say I'd be tempted to rely only on Postgres `NOW()`. If you need to pull any times into MAS for some reason, maybe do it all as relative times? (Basically avoiding using the MAS host's wall clock at all..).

Maybe this is too paranoid, perhaps? But it feels like something that  could give someone a bad time and it's not as though misconfigured clocks is a rare unheard of problem :-)...

Reply:
I've used `NOW()` only for the `expires_at` field, as this is really the time-sensitive one now. The 'dead worker shutdown' logic has a 2min threshold, which I feel like is fine to assume with have less than a minute of clock skew, especially since we're already relying on local clocks for token expirations, etc., so other things will break with large clock skews anyway

Reply:
2 mins is still a pretty tight bound/assumption on clock skew, IMO — it wouldn't surprise me to get machines that are hours out of sync in some circumstances.
You say it breaks access tokens anyway and that's true. But I think it's worth considering what the failure mode of the failure is. For access tokens, this is just going to be a wrong decision on whether a token is expired or not.

For the lock system intended to uniquely choose a worker, the failure mode is probably a lot worse than that indeed, so good to use `NOW()` there.

For this `last_seen_at`..... if this means we start assuming workers are dead when they aren't, won't this mean we potentially 'kill off' workers because we have the wrong clock? That still sounds like a hazard to me, but I haven't yet read enough to know what the effect is.

If a non-leader can kill off the leader, that certainly sounds like a hazard

Reply:
It's the leader doing the shutting down of dead workers, so the leader should in theory never be remote shutdown like that

We can probably make that a little less tight, let's say 10 minutes? In most cases anyway, workers should really gracefully shutdown, and more than a few minutes of clock skew has implications in other places, like the `id_token` generation sets the 'issued at' to now, and most OIDC libraries will validate that with max a 5min clock skew.